provider/cmd/app/app.go

```diff
@@ -408,48 +432,76 @@ func runGRPCServer(ctx context.Context, setupLog logr.Logger, log logr.Logger, s
 			commongrpc.InjectLogger(log.WithName("iri-server")),
 			commongrpc.LogRequest,
 		),
+		grpc.ConnectionTimeout(opts.ConnectionTimeout),
 	)
 	iri.RegisterMachineRuntimeServer(grpcSrv, srv)
 
-	setupLog.V(1).Info("Start listening on unix socket", "Address", opts.Address)
-	l, err := net.Listen("unix", opts.Address)
+	setupLog.V(1).Info("Start listening on unix socket", "Address", opts.Addr)
+	l, err := net.Listen("unix", opts.Addr)
 	if err != nil {
 		return fmt.Errorf("failed to listen: %w", err)
 	}
 
-	setupLog.Info("Starting grpc server", "Address", l.Addr().String())
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		<-ctx.Done()
 		setupLog.Info("Shutting down grpc server")
 		grpcSrv.GracefulStop()
-		setupLog.Info("Shut down grpc server")
+		setupLog.Info("GRPC server is shutdown")
 	}()
+
+	setupLog.Info("Starting grpc server", "Address", l.Addr().String())
 	if err := grpcSrv.Serve(l); err != nil {
 		return fmt.Errorf("error serving grpc: %w", err)
 	}
+
+	setupLog.Info("GRPC server stopped serving requests")
+
+	wg.Wait()
+
 	return nil
 }
 
-func runStreamingServer(ctx context.Context, setupLog, log logr.Logger, srv *server.Server, opts Options) error {
+func runStreamingServer(ctx context.Context, setupLog, log logr.Logger, srv *server.Server, opts HTTPServerOptions) error {
 	httpHandler := providerhttp.NewHandler(srv, providerhttp.HandlerOptions{
 		Log: log.WithName("streaming-server"),
 	})
 
 	httpSrv := &http.Server{
-		Addr:    opts.StreamingAddress,
-		Handler: httpHandler,
+		Addr:         opts.Addr,
+		Handler:      httpHandler,
+		ReadTimeout:  opts.ReadTimeout,
+		WriteTimeout: opts.WriteTimeout,
+		IdleTimeout:  opts.IdleTimeout,
 	}
 
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		<-ctx.Done()
 		setupLog.Info("Shutting down streaming server")
-		_ = httpSrv.Close()
-		setupLog.Info("Shut down streaming server")
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), opts.GracefulTimeout)
+		defer cancel()
+
+		locErr := httpSrv.Shutdown(shutdownCtx)
```

Comment:
is shutdown safe with websocket connections?

Reply:
yes, it is.

The Shutdown method is designed to gracefully close all active connections, including WebSocket connections. It achieves this by sending a close frame to each WebSocket client, allowing them to handle the server shutdown gracefully and close their connections properly.

If we were to simply close the server without using Shutdown, active WebSocket connections would be abruptly terminated, which is not ideal. Using Shutdown ensures that WebSocket clients have the opportunity to clean up and close their connections gracefully.